metadata-ingestion/src/datahub/ingestion/source/grafana/grafana_source.py

```diff
@@ -1,131 +1,427 @@
+        self.platform = config.platform
+        self.platform_instance = self.config.platform_instance
+        self.env = self.config.env
+        self.report = GrafanaSourceReport()
+
+        self.client = GrafanaAPIClient(
+            base_url=self.config.url,
+            token=self.config.service_account_token,
+            verify_ssl=self.config.verify_ssl,
+            page_size=self.config.page_size,
+            report=self.report,
+        )
+
+        # Initialize lineage extractor with graph
+        self.lineage_extractor = LineageExtractor(
+            platform=self.config.platform,
+            platform_instance=self.config.platform_instance,
+            env=self.config.env,
+            connection_to_platform_map=self.config.connection_to_platform_map,
+            graph=self.ctx.graph,
+            report=self.report,
+        )
 
     @classmethod
-    def create(cls, config_dict, ctx):
+    def create(cls, config_dict: dict, ctx: PipelineContext) -> "GrafanaSource":
         config = GrafanaSourceConfig.parse_obj(config_dict)
         return cls(config, ctx)
 
     def get_workunit_processors(self) -> List[Optional[MetadataWorkUnitProcessor]]:
-        return [
-            *super().get_workunit_processors(),
+        processors = super().get_workunit_processors()
+        processors.append(
             StaleEntityRemovalHandler.create(
-                self, self.source_config, self.ctx
-            ).workunit_processor,
-        ]
-
-    def get_report(self) -> StatefulIngestionReport:
-        return self.report
+                self, self.config, self.ctx
+            ).workunit_processor
+        )
+        return processors
 
     def get_workunits_internal(self) -> Iterable[MetadataWorkUnit]:
-        headers = {
-            "Authorization": f"Bearer {self.source_config.service_account_token.get_secret_value()}",
-            "Content-Type": "application/json",
-        }
-        try:
-            response = requests.get(
-                f"{self.source_config.url}/api/search", headers=headers
+        """Generate metadata work units"""
+        # Process folders
+        for folder in self.client.get_folders():
+            self.report.report_folder_scanned()
+            yield from self._process_folder(folder)
+
+        # Process dashboards and their panels
+        dashboards = self.client.get_dashboards()
+        for dashboard in dashboards:
+            self.report.report_dashboard_scanned()
+            yield from self._process_dashboard(dashboard)
+
+    def _process_folder(self, folder: Folder) -> Iterable[MetadataWorkUnit]:
+        """Process Grafana folder metadata"""
+        folder_key = FolderKey(
+            platform=self.config.platform,
+            instance=self.config.platform_instance,
+            folder_id=folder.id,
+        )
+
+        yield from gen_containers(
+            container_key=folder_key,
+            name=folder.title,
+            sub_types=["Folder"],
```

Comment:
If I'm not wrong, all subtype values are centralized here https://github.com/datahub-project/datahub/blob/master/metadata-ingestion/src/datahub/ingestion/source/common/subtypes.py